light-4j/0017-jdk11-branch-release.md

```diff
@@ -0,0 +1,34 @@
+### Summary
+
+The light platform 1.5.x is built on top of JDK 8 and it is production ready and we will support it until Java 8 is not supported. After JDK 11 is release, we have another jdk11 branch that contains the code on top of JDK 11. This branch will be the 2.0.X release and any new feature and defect fixes will be merged to two branches develop and jdk11 at the same time. After the release for JDK8, the code will be merged from develop to master. For JDK11, the code will be merged from jdk11 to master-jdk11. Once majority of our customers moves to the JDK11, we are going to switch the develop and master branch to JDK 11 and JDK8 will become jdk8 and master-jdk8 for maintance. 
+
+Now, for a period of time, we need to have two releases from to diferent branches. That means we have to change our release process to automate it. 
+
+### Motivation
+
+
+### Guide-level explanation
+
+If we want to automate the release process, we need to update the light-bot to support the flexibility. There are two options: enhance the light-bot or build light-bot-kotlin with DSL. I would prefer the second option as it is our long term strategy. 
+
+### Reference-level explanation
+
+
+### Drawbacks
+
+This increase the amount of the work for developers as we have to think about both branches. Automation might take some workload from developers but they need to have both JDK8 and JDK11 installed on the development machine. SDKMAN is the tool that is highly recommended. 
+
+
+### Rationale and Alternatives
+
+
+
+### Unresolved questions
+
+1. Do we want to have one command line to release both jdk8 and jdk11? Or have two separate commands?
+2. When do we switch our focus to jdk11? Waiting for Redhat docker image is ready?
+3. How to merge one feature or fix to both develop and jdk11 branch? Two PRs?
+4. Do we need to make sure that there is no significant changes in jdk11 in order to have one PR that can be merged to both branches?
+
```

Comment:
Can't we have one codebase that supports both jdk8 and jdk11?

Reply:
How we can do that? Could you please elaborate? Thanks.

Reply:
I think maybe two PRs may be more feasible. Because even if the fixs or additions in the two PRs are the same. If using one PR (development based on jdk8 branch), Will there be a conflict with jdk11 branch in the merge? Maybe the version of jdk is explained when creating a temporary development branch. E.g. feat/#123-discription-jdk8

Reply:
There are several aspects related to builds (including the ones related to Java 8/11) which we can abstract within the light-bot.

I do support the idea of moving to a Kotlin-based bot.

Reply:
Please correct me if I am wrong, but as far as I understand, jdk8 vs jdk11 is an issue during compilation, not during runtime. So if `light-4j` is compiled on jdk11 but doesn't require any special features that aren't available in java 1.8 runtime, then it should be perfectly fine for a project using jdk8 to depend on `light-4j` that has been compiled with jdk11

Reply:
But there are runtime differences between jdk8 and jdk11. For example, jdk 8 does not support ALPN protocol while jdk 11 does. Even the code compiles fine with jdk11, it'll not work seamlessly on jdk/jre 8.
So, it's not just a compile time issue.

Reply:
Sorry I think what I said is incorrect. What I think I meant to say is that, if the codebase in compiled in jdk8 it will still run in java 11. 

Please ignore my comment. I am not familiar enough with light-4j codebase and I can only imagine the complexity and cases where compatibility would be an issue.

My comment was more from a curiosity perspective than a suggestion.

Reply:
ok. Some internal classes in jdk/jre 8 are not available anymore in jdk/jre 11. So, code compiled with jdk8 actually cannot run on jdk/jre 11 in many cases.

Reply:
When migrating to jdk11, there are a lot of changes that need to be done in order to ensure that the codebase can be compiled in jdk11, so it is impossible to keep the same codebase. The best we can do is to keep the same functionality for as long as possible so that one update can just submit two PRs.